File: Sources/Brink/CodexProvider.swift
import Foundation

/// Reads the Codex CLI's OAuth token (~/.codex/auth.json) and queries the
/// ChatGPT backend usage endpoint, as documented by the CodexBar project.
struct CodexProvider: UsageProvider {
    let id = "codex"

    private static let usageURL = URL(string: "https://chatgpt.com/backend-api/wham/usage")!

    func fetch() async -> ProviderSnapshot {
        var snap = ProviderSnapshot(id: id, name: "Codex",
                                    systemImage: "terminal",
                                    windows: [], error: nil)
        guard let auth = Self.loadAuth() else {
            return ClaudeProvider.demoSnapshot(name: "Codex", systemImage: "terminal",
                                               note: L("Codex CLI credentials not found (~/.codex/auth.json)"))
        }

        var request = URLRequest(url: Self.usageURL)
        request.httpMethod = "GET"
        request.setValue("Bearer \(auth.accessToken)", forHTTPHeaderField: "Authorization")
        if let account = auth.accountID {
            request.setValue(account, forHTTPHeaderField: "ChatGPT-Account-Id")
        }
        request.setValue("Brink/1.0", forHTTPHeaderField: "User-Agent")

        do {
            let (data, response) = try await URLSession.shared.data(for: request)
            let status = (response as? HTTPURLResponse)?.statusCode ?? 0
            guard status == 200 else {
                snap.error = status == 401
                    ? L("Unauthorized — run `codex` once to refresh login")
                    : L("HTTP %d", status)
                return snap
            }
            snap.windows = Self.parseUsage(data)
            snap.updatedAt = Date()
            if snap.windows.isEmpty { snap.error = L("No usage data in response") }
            return snap
        } catch {
            snap.error = error.localizedDescription
            return snap
        }
    }

    // MARK: Auth

    struct Auth {
        var accessToken: String
        var accountID: String?
    }

    static func loadAuth() -> Auth? {
        let home = ProcessInfo.processInfo.environment["CODEX_HOME"].map { URL(fileURLWithPath: $0) }
            ?? FileManager.default.homeDirectoryForCurrentUser.appendingPathComponent(".codex")
        let url = home.appendingPathComponent("auth.json")
        guard let data = try? Data(contentsOf: url),
              let root = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any] else {
            return nil
        }
        let tokens = root["tokens"] as? [String: Any]
        guard let token = (tokens?["access_token"] as? String)
                ?? (root["access_token"] as? String) else {
            return nil
        }
        let account = (tokens?["account_id"] as? String) ?? (root["account_id"] as? String)
        return Auth(accessToken: token, accountID: account)
    }

    // MARK: Parsing

    /// Expected shape (fields defensively probed):
    /// { "rate_limit": { "primary_window": { "used_percent": 21, ... },
    ///                   "secondary_window": { "used_percent": 52, ... } } }
    static func parseUsage(_ data: Data) -> [UsageWindow] {
        guard let root = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any] else {
            return []
        }
        let rateLimit = (root["rate_limit"] as? [String: Any])
            ?? (root["rate_limits"] as? [String: Any])
            ?? root
        var windows: [UsageWindow] = []
        let ordered: [(String, String)] = [
            ("primary_window", "Current session"),
            ("primary", "Current session"),
            ("secondary_window", "Weekly"),
            ("secondary", "Weekly"),
        ]
        var seenLabels = Set<String>()
        for (key, label) in ordered {
            guard !seenLabels.contains(label),
                  let dict = rateLimit[key] as? [String: Any],
                  let pct = ClaudeProvider.number(dict["used_percent"]) else { continue }
            windows.append(UsageWindow(label: label, usedPercent: pct,
                                       resetsAt: Self.resetDate(dict)))
            seenLabels.insert(label)
        }
        return windows
    }

    static func resetDate(_ dict: [String: Any]) -> Date? {
        if let s = ClaudeProvider.number(dict["resets_in_seconds"]) {
            return Date().addingTimeInterval(s)
        }
        if let iso = ClaudeProvider.isoDate(dict["resets_at"]) {
            return iso
        }
        if let epoch = ClaudeProvider.number(dict["resets_at"]) {
            // Could be seconds or milliseconds since epoch.
            let seconds = epoch > 10_000_000_000 ? epoch / 1000.0 : epoch
            return Date(timeIntervalSince1970: seconds)
        }
        return nil
    }
}

File: windows/CodexProvider.cs
using System.IO;
using System.Net.Http;
using System.Text.Json;
using static Brink.L10n;

namespace Brink;

/// Reads the Codex CLI's OAuth token (%USERPROFILE%\.codex\auth.json, or
/// %CODEX_HOME%\auth.json) and queries the ChatGPT backend usage endpoint.
public class CodexProvider : IUsageProvider
{
    public string Id => "codex";

    private const string UsageUrl = "https://chatgpt.com/backend-api/wham/usage";

    private record Auth(string AccessToken, string? AccountId);

    public async Task<ProviderSnapshot> FetchAsync()
    {
        var snap = new ProviderSnapshot { Id = Id, Name = "Codex" };
        var auth = LoadAuth();
        if (auth == null)
            return ClaudeProvider.DemoSnapshot("Codex", L("Codex CLI credentials not found (~/.codex/auth.json)"));

        try
        {
            using var request = new HttpRequestMessage(HttpMethod.Get, UsageUrl);
            request.Headers.TryAddWithoutValidation("Authorization", $"Bearer {auth.AccessToken}");
            if (auth.AccountId != null)
                request.Headers.TryAddWithoutValidation("ChatGPT-Account-Id", auth.AccountId);
            request.Headers.TryAddWithoutValidation("User-Agent", "Brink/1.0");
            using var response = await ClaudeProvider.Http.SendAsync(request);
            var status = (int)response.StatusCode;
            if (status != 200)
            {
                snap.Error = status == 401
                    ? L("Unauthorized — run `codex` once to refresh login")
                    : L("HTTP %d", status);
                return snap;
            }
            var body = await response.Content.ReadAsStringAsync();
            snap.Windows = ParseUsage(body);
            snap.UpdatedAt = DateTime.Now;
            if (snap.Windows.Count == 0) snap.Error = L("No usage data in response");
            return snap;
        }
        catch (Exception e)
        {
            snap.Error = e.Message;
            return snap;
        }
    }

    // MARK: Auth

    private static Auth? LoadAuth()
    {
        var home = Environment.GetEnvironmentVariable("CODEX_HOME")
            ?? Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".codex");
        var file = Path.Combine(home, "auth.json");
        try
        {
            if (!File.Exists(file)) return null;
            using var doc = JsonDocument.Parse(File.ReadAllText(file));
            var root = doc.RootElement;
            JsonElement tokens = default;
            bool hasTokens = root.TryGetProperty("tokens", out tokens)
                && tokens.ValueKind == JsonValueKind.Object;
            var token = (hasTokens ? ClaudeProvider.GetString(tokens, "access_token") : null)
                ?? ClaudeProvider.GetString(root, "access_token");
            if (token == null) return null;
            var account = (hasTokens ? ClaudeProvider.GetString(tokens, "account_id") : null)
                ?? ClaudeProvider.GetString(root, "account_id");
            return new Auth(token, account);
        }
        catch { return null; }
    }

    // MARK: Parsing

    /// Expected shape (fields defensively probed):
    /// { "rate_limit": { "primary_window": { "used_percent": 21, ... },
    ///                   "secondary_window": { "used_percent": 52, ... } } }
    public static List<UsageWindow> ParseUsage(string data)
    {
        var windows = new List<UsageWindow>();
        try
        {
            using var doc = JsonDocument.Parse(data);
            var root = doc.RootElement;
            if (root.ValueKind != JsonValueKind.Object) return windows;
            JsonElement rateLimit = root;
            if (root.TryGetProperty("rate_limit", out var rl) && rl.ValueKind == JsonValueKind.Object)
                rateLimit = rl;
            else if (root.TryGetProperty("rate_limits", out var rls) && rls.ValueKind == JsonValueKind.Object)
                rateLimit = rls;

            (string Key, string Label)[] ordered =
            {
                ("primary_window", "Current session"),
                ("primary", "Current session"),
                ("secondary_window", "Weekly"),
                ("secondary", "Weekly"),
            };
            var seen = new HashSet<string>();
            foreach (var (key, label) in ordered)
            {
                if (seen.Contains(label)) continue;
                if (!rateLimit.TryGetProperty(key, out var dict) || dict.ValueKind != JsonValueKind.Object) continue;
                var pct = ClaudeProvider.GetNumber(dict, "used_percent");
                if (pct == null) continue;
                windows.Add(new UsageWindow { Label = label, UsedPercent = pct.Value, ResetsAt = ResetDate(dict) });
                seen.Add(label);
            }
        }
        catch { }
        return windows;
    }

    private static DateTime? ResetDate(JsonElement dict)
    {
        if (ClaudeProvider.GetNumber(dict, "resets_in_seconds") is double s)
            return DateTime.Now.AddSeconds(s);
        if (ClaudeProvider.GetIsoDate(dict, "resets_at") is DateTime iso)
            return iso;
        if (ClaudeProvider.GetNumber(dict, "resets_at") is double epoch)
        {
            // Could be seconds or milliseconds since epoch.
            var seconds = epoch > 10_000_000_000 ? epoch / 1000.0 : epoch;
            return DateTimeOffset.FromUnixTimeSeconds((long)seconds).LocalDateTime;
        }
        return null;
    }
}
